src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1531 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Management.Automation.Language;
+using System.Management.Automation.Runspaces;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using Microsoft.PowerShell.Commands;
+
+using CimClass = Microsoft.Management.Infrastructure.CimClass;
+using CimInstance = Microsoft.Management.Infrastructure.CimInstance;
+
+namespace System.Management.Automation
+{
+    /// <summary>
+    /// Enum describing permissions to use runtime evaluation during type inference
+    /// </summary>
+    [Flags]
+    public enum TypeInferenceRuntimePermissions {
+        /// <summary>
+        /// No runtime use is allowed
+        /// </summary>
+        None = 0,
+        /// <summary>
+        /// Use of SafeExprEvaluator visitor is allowed
+        /// </summary>
+        AllowSafeEval = 1,
+    }
+
+    /// <summary>
+    /// static class containting methods to work with type inference of abstract syntax trees
+    /// </summary>
+    public static class AstTypeInference
+    {
```

Comment:
I'd like to keep the api internal for now.
In the short term, we are considering adding an "experimental" option - and when we have that, then I'd like to make this public but experimental, meaning external code needs to opt-in explicitly or the api wouldn't work.